accounts/abi/bind/backends/simulated.go

```diff
@@ -0,0 +1,184 @@
+
+	b.pendingBlock = blocks[0]
+	b.pendingState, _ = state.New(b.pendingBlock.Root(), b.database)
+}
+
+// ContractCall implements ContractCaller.ContractCall, executing the specified
+// contract with the given input data.
+func (b *SimulatedBackend) ContractCall(contract common.Address, data []byte, pending bool) ([]byte, error) {
+	// Create a copy of the current state db to screw around with
+	var (
+		block   *types.Block
+		statedb *state.StateDB
+	)
+	if pending {
+		block, statedb = b.pendingBlock, b.pendingState
+	} else {
+		block = b.blockchain.CurrentBlock()
+		statedb, _ = b.blockchain.State()
+	}
+	statedb = statedb.Copy()
+
+	// Set infinite balance to the a fake caller account
+	from := statedb.GetOrNewStateObject(common.Address{})
+	from.SetBalance(common.MaxBig)
+
+	// Assemble the call invocation to measure the gas usage
+	msg := callmsg{
+		from:     from,
+		to:       &contract,
+		gasPrice: new(big.Int),
+		gasLimit: common.MaxBig,
+		value:    new(big.Int),
+		data:     data,
+	}
+	// Execute the call and return
+	vmenv := core.NewEnv(statedb, b.blockchain, msg, block.Header())
+	gaspool := new(core.GasPool).AddGas(common.MaxBig)
+
+	out, _, err := core.ApplyMessage(vmenv, msg, gaspool)
+	return out, err
+}
+
+// AccountNonce implements ContractTransactor.AccountNonce, retrieving the nonce
+// currently pending for the account.
+func (b *SimulatedBackend) AccountNonce(account common.Address) (uint64, error) {
+	return b.pendingState.GetOrNewStateObject(account).Nonce(), nil
+}
+
+// GasPrice implements ContractTransactor.GasPrice. Since the simulated chain
+// doens't have miners, we just return a gas price of 1 for any call.
+func (b *SimulatedBackend) GasPrice() (*big.Int, error) {
+	return big.NewInt(1), nil
+}
+
+// GasLimit implements ContractTransactor.GasLimit, executing the requested code
+// against the currently pending block/state and returning the used gas.
+func (b *SimulatedBackend) GasLimit(sender common.Address, contract *common.Address, value *big.Int, data []byte) (*big.Int, error) {
+	// Create a copy of the currently pending state db to screw around with
+	var (
+		block   = b.pendingBlock
+		statedb = b.pendingState.Copy()
+	)
+
+	// Set infinite balance to the a fake caller account
+	from := statedb.GetOrNewStateObject(sender)
+	from.SetBalance(common.MaxBig)
+
+	// Assemble the call invocation to measure the gas usage
+	msg := callmsg{
+		from:     from,
+		to:       contract,
+		gasPrice: new(big.Int),
+		gasLimit: common.MaxBig,
+		value:    value,
+		data:     data,
+	}
+	// Execute the call and return
+	vmenv := core.NewEnv(statedb, b.blockchain, msg, block.Header())
+	gaspool := new(core.GasPool).AddGas(common.MaxBig)
+
+	_, gas, err := core.ApplyMessage(vmenv, msg, gaspool)
+	return gas, err
+}
+
+// Transact implements ContractTransactor.SendTransaction, delegating the raw
+// transaction injection to the remote node.
+func (b *SimulatedBackend) SendTransaction(tx *types.Transaction) error {
```

Comment:
The reason I did it like this is because I reused `core.ChainMaker`, which already implements almost all of the block generation and transaction processing logic. However, since it pregenerates the blocks and doesn't support further modifying them, my choices were to do a bit of extra work (which given that it's a tester, shouldn't matter that much), or on the other hand duplicate almost all the tx processing and block assembly code yet again here too.